### ISSUE TITLE
Support activity functions with variadic arguments  

### ISSUE THREAD

**Issue by aryzhov** [temporalio/sdk-go#1114]
**Is your feature request related to a problem? Please describe.**

I need to make gRPC function calls from my workflows. All gRPC client functions generated by `protoc` have the variadic  parameter `grpc.CallOption`:

```protobuf
type SomeClient interface {
	SomeMethod(ctx context.Context, in *SomeRequest, opts ...grpc.CallOption) (*SomeResponse, error)
}
```

I am able to register `SomeMethod` as an activity but when I try calling it from a workflow, passing an empty array as the last argument, I get an error:

```
reflect: cannot use []grpc.CallOption as type grpc.CallOption in Call
```

I tried different ways of passing the `opts` parameter (omitting, null, a single instance of `grpc.CallOption`) but neither of these ways have worked. 

The issue is generic and applies to all variadic functions regardless of parameter type. I was able to reproduce the problem by modifying `GreetingSample` and adding a variadic argument to the `GetGreeting` activity.

**Describe the solution you'd like**

I can see two ways of implementing this feature:

1. Pass variadic args as an array to `ExecuteActivity()`. 
2. Pass variadic args as individual parameters to `ExecuteActivity()`. 

Upon examining source code, The first option needs a change to `executeFunction()` function in `internal_worker.go` to unpack the array. The second option requires a change to `decodeArgsToRawValues()`, also in `internal_worker.go`. I am not sure which option is better. 

**Describe alternatives you've considered**

I was able to implement a workaround by creating a utility function:
```golang
func wrapGrpcActivity[P any, R any](f func(ctx context.Context, req *P, opts ...grpc.CallOption) (*R, error)) func(ctx context.Context, req *P, opts []grpc.CallOption) (*R, error) {
	return func(ctx context.Context, req *P, opts []grpc.CallOption) (*R, error) {
		return f(ctx, req, opts...)
	}
}
```

Then I use the function when registering gRPC methods:
```golang
worker.RegisterActivityWithOptions(wrapGrpcActivity(someClient.SomeMethod), activity.RegisterOptions{Name: "SomeMethod"})
```

and I can invoke the activity as follows:

```golang
err := workflow.ExecuteActivity(ctx, "SomeMethod", &some.SomeRequest{}, nil).Get(ctx, &resp)
```

**Additional context**

I would like to add that supporting variadic args would allow developers more easily integrate existing third party APIs as activities into workflows without additional wrapping.

**Comment by cretz** [comment]
Variadic arguments are supported I believe if you pass the slice instead of just a single value. I think this a reasonable compromise currently.

Your code will not work anyways. `grpc.CallOption` is not a convertible/serializable value. You should use a wrapper, you cannot expose gRPC client calls directly with their non-serializable arugments as activities.

**Comment by aryzhov** [comment]
@cretz Let me answer the two points you made separately. First, as I had indicated, this is a generic issue, not specific to gRPC. I was able to reproduce it with an example GreetingService. If I add a variadic parameter:
```golang
func (a *Activities) GetGreeting(p ...string) (string, error) {
```
and then change `GreetingSample` to pass this parameter:
```golang
err := workflow.ExecuteActivity(ctx, a.GetGreeting, []string{"a", "b"}).Get(ctx, &greetResult)
```
I get an error:
```
reflect: cannot use []string as type string in Call
```
There is a stack trace:
```
reflect.Value.call({0x103142400?, 0x140002a05c0?, 0x12af355e8?}, {0x102f8ba78, 0x4}, {0x140000a1428, 0x1, 0x0?})
        /opt/homebrew/opt/go/libexec/src/reflect/value.go:453 +0x13c0
reflect.Value.Call({0x103142400?, 0x140002a05c0?, 0x14000447778?}, {0x140000a1428?, 0x140000a4310?, 0x9?})
        /opt/homebrew/opt/go/libexec/src/reflect/value.go:370 +0x90
go.temporal.io/sdk/internal.executeFunction({0x103142400, 0x140002a05c0}, {0x140001e8950, 0x1, 0x103275101?})
        /Users/alex/go/pkg/mod/go.temporal.io/sdk@v1.22.2/internal/internal_worker.go:1796 +0x318
go.temporal.io/sdk/internal.executeFunctionWithContext({0x1032751b8?, 0x140004403c0}, {0x103142400, 0x140002a05c0}, {0x140001e8950, 0x1, 0x1})
        /Users/alex/go/pkg/mod/go.temporal.io/sdk@v1.22.2/internal/internal_worker.go:1780 +0x160
go.temporal.io/sdk/internal.(*activityEnvironmentInterceptor).ExecuteActivity(0x140000b6a80, {0x103275180?, 0x14000440360?}, 0x140000a1410)
```

**Comment by aryzhov** [comment]
@cretz For your second point, I agree that `grpc.CallOption` is not convertible, and I would need a wrapper if I had wanted to use it. However, I don't want to pass any call options, and in many cases these call options aren't being used, in which case convertibility isn't an issue.

**Comment by cretz** [comment]
:+1: Yeah we call https://pkg.go.dev/reflect#Value.Call not https://pkg.go.dev/reflect#Value.CallSlice. I assume `workflow.ExecuteActivity(ctx, a.GetGreeting, "a", "b").Get(ctx, &greetResult)` also doesn't work. I think we can support variadic args, but I think for compatibility it may have to be with how `Call` does it (so making what I put there work).

I wouldn't count on non-convertible optional variadic options always being safe. We may have static analyzers or other things in the future that check entire signatures eagerly (e.g. at registration).

**Comment by aryzhov** [comment]
Yes, the second form (`err := workflow.ExecuteActivity(ctx, a.GetGreeting, "a", "b").Get(ctx, &greetResult)`) also doesn't work. It fails with a different error:
```
Error unable to decode the activity function input payload with error: payload item 0: unable to decode: json: cannot unmarshal string into Go value of type []string for function name: GetGreeting
```

**Comment by mashail** [comment]
I support @aryzhov in his quest.
